lib/internal/errors.js

```diff
@@ -0,0 +1,139 @@
+'use strict';
+
+// The whole point behind this internal module is to allow Node.js to no
+// longer be forced to treat every error message change as a semver-major
+// change. The NodeError classes here all expose a `code` property whose
+// value statically and permanently identifies the error. While the error
+// message may change, the code should not.
+
+const kCode = Symbol('code');
+const kSymbolsMap = new Map();
+
+function symbolFor(name) {
+  var sym = kSymbolsMap.get(name);
+  if (sym === undefined) {
+    sym = Symbol(name);
+    kSymbolsMap.set(name, sym);
+  }
+  return sym;
+}
+
+var assert, util;
+function lazyAssert() {
+  if (!assert)
+    assert = require('assert');
+  return assert;
+}
+
+function lazyUtil() {
+  if (!util)
+    util = require('util');
+  return util;
+}
+
+class NodeError extends Error {
+  constructor(key /**, ...args **/) {
+    lazyAssert().strictEqual(typeof key, 'string');
+    const args = Array(arguments.length - 1);
+    for (var n = 1; n < arguments.length; n++)
+      args[n - 1] = arguments[n];
+    super(message(key, args));
+    this[kCode] = key;
+    Error.captureStackTrace(this, NodeError);
+  }
+
+  get name() {
+    return `Error[${this[kCode]}]`;
+  }
+
+  get code() {
+    return this[kCode];
+  }
+}
+
+class NodeTypeError extends TypeError {
+  constructor(key /**, ...args **/) {
+    lazyAssert().strictEqual(typeof key, 'string');
+    const args = Array(arguments.length - 1);
+    for (var n = 1; n < arguments.length; n++)
+      args[n - 1] = arguments[n];
+    super(message(key, args));
+    this[kCode] = key;
+    Error.captureStackTrace(this, NodeTypeError);
+  }
+
+  get name() {
+    return `TypeError[${this[kCode]}]`;
+  }
+
+  get code() {
+    return this[kCode];
+  }
+}
+
+class NodeRangeError extends RangeError {
+  constructor(key /**, ...args **/) {
+    lazyAssert().strictEqual(typeof key, 'string');
+    const args = Array(arguments.length - 1);
+    for (var n = 1; n < arguments.length; n++)
+      args[n - 1] = arguments[n];
+    super(message(key, args));
+    this[kCode] = key;
+    Error.captureStackTrace(this, NodeRangeError);
+  }
+
+  get name() {
+    return `RangeError[${this[kCode]}]`;
+  }
+
+  get code() {
+    return this[kCode];
+  }
+}
+
+function message(key, args) {
+  const assert = lazyAssert();
+  const util = lazyUtil();
+  const msg = exports[symbolFor(key)];
```

Comment:
If I am reading this right, the second time a key is passed into this, it will be in the map and this assertion won't fail, even though the error is not properly registered with `E`?